src/DeviceInterface/InterfacePrototype.js

```diff
@@ -537,9 +513,87 @@ class InterfacePrototype {
     openManagePasswords () {}
     storeFormData (_values) {}
 
-    /** @param {FeatureToggleNames} _name */
-    supportsFeature (_name) {
-        return false
+    /** @param {FeatureToggleNames} name */
+    supportsFeature (name) {
+        return this.supportedFeatures.includes(name)
+    }
+
+    /**
+     * `preAttachTooltip` happens just before a tooltip is show - features may want to append some data
+     * at this point.
+     *
+     * For example, if password generation is enabled, this will generate
+     * a password and send it to the tooltip as though it were a stored credential.
+     *
+     * @param {TopContextData} topContextData
+     * @param {HTMLInputElement} input
+     * @param {{isSignup: boolean|null}} form
+     */
+    preAttachTooltip (topContextData, input, form) {
```

Comment:
Unrelated to this, but the more time passes the more I think that having the same code run in the page and in the tooltip is a genuinely bad idea. It was a shortcut we took at the start of the top autofill project and it's created all sorts of problems. We'll have to separate the two branches out at some point.